### DISCUSSION TITLE
HOWTO load fibo into Stardog

### DISCUSSION THREAD

**Discussion by bblfish** [edmcouncil/fibo#1370] [General]
I wanted to find out how OWL DL reasoning on FIBO worked with [Stardog](https://www.stardog.com/) as I was having trouble getting the reasoners in Protégé to finish in a reasonable time.

Stardog does not seem to go through the [FIDO git repository](https://github.com/edmcouncil/fibo/) file by file when one points it to the [AboutFIBOProd.rdf](https://github.com/edmcouncil/fibo/blob/master/AboutFIBOProd.rdf) as Protégé does, so I tried to zip the github directory and run the command, with strict parsing set to false

```zsh
> stardog-admin db create -o strict.parsing=false -n fibo ../fibo.zip
```

But this loads only 7795 triples only as shown by the query, and gives an error message.

```SPARQL
SELECT (count(?A) as ?size)
WHERE { ?A ?R ?B }
```

I am expecting more (though I can't quite tell in Protege how many triples it loads, as its statics don't make that clear, but it says there are 85000 rules). Protégé hangs for a while  on my machine when I try to open the SPARQL tab so I could not run the same query immediately. After a while I found it working and running the above SPARQL query I got 422551 as a result. 
So I am not sure what the correct amount of triples should be here.

Then it occurred to me that in any case the github version contains both the development and the production version. So I downloaded the turtle production version.

```zsh
> stardog-admin db create -o strict.parsing=false -n fibo fibo-prod.ttl.zip
```

which loaded 86688 triples, running the same query. 
I wonder what the right number of triples loaded should be.

It did show this error

```
Errors were encountered during loading:
/Volumes/Dev/Programming/RDF/OWL/Data/fibo-prod.ttl.zip: IRI included an unencoded space: '32' [L1]
```

I am a bit surprised that the github version is all in RDF/XML rather than Turtle. Does that not make participation by developers more difficult? (Given that RDF/XML is notoriously difficult to read and learn)

**Comment by dallemang** [comment]
Hi Henry,

You might try the "quickstart" files, available at
https://spec.edmcouncil.org/fibo/OWL (near the bottom of the matrix).
These include all of prod/dev respectively, without any imports required.

Before we had the quickstart files (which might work in stardog directly),
we published instructions for loading into Stardog; these instructions
should work even with newer versions of stardog.  You can find them at
https://spec.edmcouncil.org/fibo/ontology-tools  (look for Stardog).

I would be interested to know what methods work best for you, and update
our docs accordingly if you learn anything new.


Dean

**Comment by bblfish** [comment]
Thanks @dallemang . 
I can't quite tell whether the problem is  on Stardogs or your side, but I guess it is more likely to be on stardogs side. I pinged them in an issue on [their stardog-tutorial github repo](https://github.com/stardog-union/stardog-tutorials/issues/6) but I don't know if that is the best way to make them aware of the problem. If someone knows a better channel  let me know please.

I interpreted the recipe for stardog as telling me to do the following on the production version of FIBO.

```zsh
unzip fibo-prod.rdf.zip
cd fibo/ontology/master/2020Q1                   # there is no latest dir, only a 2020Q1 dir
find . -name "About*.rdf"                                # could not find any such files 
find . -name "catalog*" -exec rm {} \;            # deleted over 10
find . -name "*.rdf" -exec sed -i .old "s+/fibo/ontology/master/2020Q1/+/fibo/ontology/+" {} \;
find . -name "*.rdf.old" -exec grep -q "/fibo/ontology/master/2020Q1" {} \; -print | wc
find . -name "*.rdf" -exec grep -q "/fibo/ontology/master/2020Q1" {} \; -print | wc
find . -name "*.rdf.old" -exec rm {} \;             # the changes worked so delete them 
cd ../..
mv master/2020Q1/* .
cd ../..
zip -r Fibo4Star.rdf.zip fibo
```

As the comments indicate there were a few things that did not quite match, such as I could not find any `About*` files. So I tried this again on the development release.

```zsh
curl -o fibo-dev.rdf.zip https://spec.edmcouncil.org/fibo/ontology/master/latest/dev.rdf.zip
unzip fibo-dev.rdf.zip
cd fibo/ontology/master/latest
find . -name "About*.rdf"
find . -name "catalog*"
find . -name "catalog*" -exec rm {} \;
find . -name "*.rdf" -exec sed -i .old "s+/fibo/ontology/master/latest/+/fibo/ontology/+" {} \;
find . -name "*.rdf" -exec diff {} {}.old \; | less
find . -name "*.rdf.old" -exec rm {} \;
cd ../..
mv master/latest/* .
cd ../..
zip -r Fibo4Star-dev.rdf.zip fibo
stardog-admin db create -n fibo Fibo4Star-dev.rdf.zip
```

Yet I get the following error when loading (note there are no About files here either):

```zsh
> stardog-admin db create -n fibo Fibo4Star-dev.rdf.zip
Java HotSpot(TM) 64-Bit Server VM warning: Option UseParallelOldGC was deprecated in version 14.0 and will likely be removed in a future release.
Bulk loading data to new database fibo.
Errors were encountered during loading:
/Volumes/Dev/Programming/RDF/OWL/Data/Fibo4Star-dev.rdf.zip: org.xml.sax.SAXParseException; Premature end of file. [L1]
Loaded 0 triples to fibo from 1 file(s) in 00:00:00.293 @ 0.0K triples/sec.
Successfully created database 'fibo'.
```

If I load it from the User Interface `Stardog Studio` I get an error in their log files:

```java
WARN  2020-06-11 18:08:37,764 [stardog-user-5] com.complexible.stardog.protocols.http.server.StardogUndertowErrorHandler:accept(64): Unexpected exception was handled by the server
com.complexible.tx.api.HeuristicRollbackException: There was a fatal failure during preparation of a99d9a8c-5e9f-44cd-90b7-d7b730ac3670 com.stardog.stark.io.InvalidRDF: unqualified attribute 'prefer' not allowed [L3:78]
        at com.complexible.tx.api.impl.DefaultTransaction.runPreparePhase(DefaultTransaction.java:426) ~[stardog-7.3.1.jar:?]
        at com.complexible.tx.api.impl.DefaultTransaction.commit(DefaultTransaction.java:308) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.db.DatabaseConnectionImpl.commit(DatabaseConnectionImpl.java:493) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.db.DelegatingDatabaseConnection.commit(DelegatingDatabaseConnection.java:290) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.db.DatabaseImpl$MonitoredDbConnection.commit(DatabaseImpl.java:1632) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.db.DatabaseImpl$DBConnectionWrapper.commit(DatabaseImpl.java:1383) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.db.DelegatingDatabaseConnection.commit(DelegatingDatabaseConnection.java:290) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.StardogKernel$KernelDbConnection.commit(StardogKernel.java:3107) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.db.DelegatingDatabaseConnection.commit(DelegatingDatabaseConnection.java:290) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.StardogKernel$NotifyingDatabaseConnection.commit(StardogKernel.java:3382) ~[stardog-7.3.1.jar:?]
        at com.complexible.stardog.protocols.http.server.ProtocolUtils.executeCommit(ProtocolUtils.java:749) ~[stardog-protocols-http-server-7.3.1.jar:?]
        at com.complexible.stardog.protocols.http.server.TransactionService.commit(TransactionService.java:136) ~[stardog-protocols-http-server-7.3.1.jar:?]
        at com.stardog.http.server.undertow.jaxrs.ExtractRoutes.lambda$handleIt$5(ExtractRoutes.java:192) ~[stardog-protocols-http-server-7.3.1.jar:?]
        at org.apache.shiro.subject.support.SubjectRunnable.doRun(SubjectRunnable.java:120) [shiro-core-1.3.0.jar:1.3.0]
        at org.apache.shiro.subject.support.SubjectRunnable.run(SubjectRunnable.java:108) [shiro-core-1.3.0.jar:1.3.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
(END)
```

Here are the two files I put together:
* [Fibo4Star.rdf.zip](https://github.com/edmcouncil/fibo/files/4765993/Fibo4Star.rdf.zip)
* [Fibo4Star-dev.rdf.zip](https://github.com/edmcouncil/fibo/files/4765985/Fibo4Star-dev.rdf.zip)

**Comment by bblfish** [comment]
Ok, I found that loading the fibo data one file at a time works. 
After doing all the operations above to rename the directories then going 
below the `fibo` root I can do the following:

```bash
find fibo -name "*.rdf" -exec stardog data add fibo {} \;
```

This is slow but it works. From within Stardog Studio I count the number of triples to be 119925 using the SPARQL query shown above.

I wonder if one still needs changing the URLs in the files and the diretory renaming process shown earlier with this method. Something to investigate will be if one can load multiple files or even all the files using `xargs` in one go using this method.

**Comment by bblfish** [comment]
I turned on OWL-DL inferencing and the DB got stuck, so I had to kill it, which removed the DB. That gave me the opportunity to try the following command from the root of the git repo

```bash
find */ -name "*.rdf" | xargs stardog data add fibo
```

which added all files in one go in 2 seconds.
(As I did not do any directory renamings there might be a problem with this method, which I will soon discover I guess)
